eval/topdown.go

```diff
@@ -0,0 +1,786 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package eval
+
+import "github.com/open-policy-agent/opa/opalog"
+import "fmt"
+
+var _ = fmt.Println
+
+// Bindings is the structure used to ground expressions.
+type Bindings map[opalog.Var]opalog.Value
+
+// TopDownContext contains the state of the evaluation process.
+//
+// TODO(tsandall): profile perf/memory usage with current approach;
+// the current approach copies the TopDownContext structure for each
+// step and binding. This avoids the need to undo steps and bindings
+// each time the proof fails but this may be too expensive.
+type TopDownContext struct {
+	Rule     *opalog.Rule
+	Bindings Bindings
+	Index    int
+	Previous *TopDownContext
+	Store    Storage
+}
+
+// Current returns the current expression to evaluate.
+func (ctx *TopDownContext) Current() *opalog.Expr {
+	return ctx.Rule.Body[ctx.Index]
+}
+
+// Step returns a new context to evaluate the next expression.
+func (ctx *TopDownContext) Step() *TopDownContext {
+	return &TopDownContext{
+		Rule:     ctx.Rule,
```

Comment:
Yes, I think we could rename this to Query and probably just accept a slice of expressions (or an `opalog.Body`). This would probably make TopDown easier to use as a library (otherwise callers could have to construct a fake rule in order to run the algorithm.)